📖 Revamp "Watching Resources" documentation for accuracy and clarifty (#4170)
kubernetes-sigs/kubebuilder: docs/book/src/reference/watching-resources.md

project but one that you need to watch and respond to.
For example, if **Operator A** manages a `MyApp` CRD for application deployment,
and **Operator B** handles backups, **Operator B** can watch the `MyApp` CRD as an external type
to trigger backup operations based on changes in `MyApp`.

In this scenario, **Operator B** could define a `BackupConfig` CRD that relies on the state of `MyApp`.
By treating `MyApp` as a **Secondary Resource**, **Operator B** can watch and reconcile changes in **Operator A**'s `MyApp`,
ensuring that backup processes are initiated whenever `MyApp` is updated or scaled.

## General Concept of Watching Resources

Whether a resource is defined within your project or comes from an external project, the concept of **Primary**
and **Secondary Resources** remains the same:
- The **Primary Resource** is the resource the controller is primarily responsible for managing.
- **Secondary Resources** are those that are required to ensure the primary resource works as desired.

Therefore, regardless of whether the resource was defined by your project or by another project,
your controller can watch, reconcile, and manage changes to these resources as needed.

> ## Why watch the secondary resources matters?

Reviewer: There is a grammatical error here.

```suggestion
## Why does watching the secondary resources matter?
```

```suggestion
## Why watching the secondary resources matters
```

```suggestion
## Why watch the secondary resources?
```

When building a Kubernetes controller, it’s crucial to not only focus
on **Primary Resources** but also to monitor **Secondary Resources**.
Failing to track these resources can lead to inconsistencies in your
controller's behavior and the overall cluster state.

Secondary resources may not be directly managed by your controller,
but changes to these resources can still significantly
impact the primary resource and your controller's functionality.
Here are the key reasons why it's important to watch them:

- **Ensuring Consistency**:

>     - Secondary resources (e.g., child objects or external dependencies) may diverge from their expected state.

Reviewer: ```suggestion
    - Secondary resources (e.g., child objects or external dependencies) may diverge from their last seen state.
```

... or "last reconciled state" or "last known state" — the state is not expected

Author: the righ technical common term is desired state. I will change for that

    For instance, a secondary resource may be modified or deleted, causing the system to fall out of sync.
    - Watching secondary resources ensures that any changes are detected immediately, allowing the controller to

>     reconcile and restore the desired state.

Reviewer: ```suggestion
    reconcile and restore consistency.
```

Author: The  the desired state is the common. technical term so if you do not mind I will keep this one.

if err := ctrl.NewControllerManagedBy(mgr).
   For(&<YourAPISpec>{}). <-- See there that the Controller is For this API
   Complete(r); err != nil {
   return err
}
```

## Secondary Resources

Your controller may also need to manage **Secondary Resources**,
which are resources created and owned by the controller to support
the **Primary Resource**.

These secondary resources, such as `Services`, `ConfigMaps`, or `Deployments`,
are tied to the Primary Resource via [OwnerReferences][owner-ref-k8s-docs].
Changes to these secondary resources can directly impact the **Primary Resource**,
so the controller must watch and reconcile these resources accordingly.

Additionally, if the **Primary Resource** is deleted, Kubernetes' garbage collection
mechanism ensures that all associated **Secondary Resources** are automatically deleted in a cascading manner.

<aside class="note">
<h1> Watching External Types (API/CRD Defined by Another Operator) </h1>

In Kubebuilder, resources can be categorized into two types:
- **Core Types**: API types defined by Kubernetes itself, such as `Pods`, `Services`, and `Deployments`.

> - **External Types**: API types defined in other projects, such as CRDs managed by another operator.

Author: Primary and Secondary are the technical terms adopt across the community  so we need to stick with them
Then Core and External types are the kubebuilder terms to explain the k8s kinds which are registered by default and the third-part/external ones. 

I am not sure if I followed all your comments but I will re-read the text here and see.

> - **External Types**: API types defined in other projects, such as CRDs managed by another operator.

Author: Revamp all so I am closing this one but please feel free to check if you see that it requires further improvements,
